### DISCUSSION TITLE
Access the `endpoint` from `http::Request` in a custom Connector

### DISCUSSION THREAD

**Discussion by gagbo** [awslabs/aws-sdk-rust#679] [Q&A] [answered]
Hello,

I'm currently implementing my own connector adapter to use the SDK in a wasm binary with custom bindings:
```rust
// Adapter is an empty struct. (I tried to use connector_fn but had that function future freezing, never woken up by my runtime somehow)
impl Service<http::Request<SdkBody>> for Adapter {
    type Response = http::Response<SdkBody>;

    type Error = ConnectorError;

    #[allow(clippy::type_complexity)]
    type Future = std::pin::Pin<
        Box<dyn std::future::Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>,
    >;

    fn poll_ready(
        &mut self,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, req: http::Request<SdkBody>) -> Self::Future {
        log(format!(
            "CloudWatch: adapter: sending request to {}",
            req.uri()
        ));
        // The specific bindings call that is a future
        let fut = make_http_request(self.try_into_bindings_request(req).unwrap());
        Box::pin(async move {
            Ok(http::Response::new(SdkBody::from(
                fut.await.map_err(downcast_error)?.body,
            )))
        })
    }
}
```

Everything works fine, except that the request `req` doesn't contain the endpoint I want to hit for some reason. `req.uri()` is just `/` for a request to list metrics in cloudwatch:

```console
Request { method: POST, uri: /, version: HTTP/1.1, headers: {"content-type": "application/x-www-form-urlencoded", "content-length": "94"}, body: SdkBody { inner: Once(Some(b"Action=ListMetrics&Version=2010-08-01&Namespace=AWS%2FRDS&MetricName=MaximumUsedTransactionIDs")), retryable: true } }
```
is what I'm getting as `req` basically (with some parameters added to the query). How would I get back the base URL to use to make the request?

**Comment by gagbo** [comment]
From inspecting the `SdkConfig` that I build, it seems that I don't have any endpoint resolver setup. I only hard code the credentials, and then insert the region in the SdkConfiguration I'm building. I assumed that passing the region would be enough to have a default resolver that would work.

```rust
// Building a SdkConfig from the wasm app configuration:
impl From<&Config> for SdkConfig {
    fn from(conf: &Config) -> Self {
        SdkConfig::builder()
            .region(Region::new(conf.region.clone()))
            .credentials_provider(SharedCredentialsProvider::new(Credentials::from_keys(
                &conf.access_key_id.0,
                &conf.secret_access_key.0,
                None,
            )))
            .build()
    }
}
```

And then this assertion fails when I call it
```rust
    let shared_config = SdkConfig::from(&config);
    shared_config
        .endpoint_resolver()
        .ok_or_else(|| Error::Invocation {
            message: "The shared config does not have an endpoint resolver!!".to_string(),
        })?;
            let config = aws_sdk_cloudwatch::config::Builder::from(&shared_config)
        .retry_config(RetryConfig::disabled())
        .build();

// ...
```

Since I'm never mentioning an override of the endpoint resolver, I was expecting to obtain a default one that works. I'm mostly trying to avoid `aws_config` because I have no environment to pull configuration from.

**Reply by jdisanti** [reply]
Currently the only way to get defaults is to use `aws_config::load_from_env` or `aws_config::from_env`. You can override where `aws_config` draws from with `aws_config::from_env` since it [has a builder pattern](https://docs.rs/aws-config/0.51.0/aws_config/struct.ConfigLoader.html), so it might work for you if you do that. If it doesn't, I would love to hear more about it to see if we can better support your use-case.

**Reply by gagbo** [reply]
Hello!

I just tried migrating to using `from_env` but it's still failing sadly:

```rust
#[serde(rename_all = "camelCase")]
pub struct Config {
    pub region: String,
    // https://docs.rs/aws-types/0.51.0/aws_types/credentials/struct.Credentials.html#method.from_keys
    pub access_key_id: Secret,
    pub secret_access_key: Secret,
}

impl Config {
    pub async fn generate_sdk_config(&self) -> SdkConfig {
        aws_config::from_env()
            .region(Region::new(self.region.clone()))
            .credentials_provider(SharedCredentialsProvider::new(Credentials::from_keys(
                &self.access_key_id.0,
                &self.secret_access_key.0,
                None,
            )))
            .load()
            .await
    }
}

/// Later at usage site
    let shared_config = config.generate_sdk_config().await;
    let config = aws_sdk_cloudwatch::config::Builder::from(&shared_config)
        .retry_config(RetryConfig::disabled())
        .build();
/// The request logged still is
// Request { method: POST, uri: /, version: HTTP/1.1, headers: {"content-type": "application/x-www-form-urlencoded", "content-length": "94"}, body: SdkBody { inner: Once(Some(b"Action=ListMetrics&Version=2010-08-01&Namespace=AWS%2FRDS&MetricName=MaximumUsedTransactionIDs")), retryable: true } }

```

Basically our use case is to bundle (part of) the SDK in a wasm binary that we want to use as a plugin to run in our custom runtimes (the runtimes used at fiberplane.com). That's why I didn't think about relying on the environment for any operation, the plugins only realistically have access to the data it's sent through exposed binding function calls. The main things that I needed to do until now are:
- Fork the SDK to use an unreleased version of `ring` dependency (to stop the SDK to try to link to `wasm-bindgen` symbols when compiled to `wasm32-unknown-unknown`)
- Create an Adapter to be used as Connector in the clients that I build in the wasm (which uses the custom wasm bindings to make the http queries through the host)

That shared SdkConfig sends the `endpoint_resolver: None` [as is to cloudwatch-specific config builder](https://github.com/awslabs/aws-sdk-rust/blob/31cfae2cf23be0c68a47357070dea1aee9227e3a/sdk/cloudwatch/src/config.rs#L358), which eventually gets [filled by the build call with a default resolver](https://github.com/awslabs/aws-sdk-rust/blob/31cfae2cf23be0c68a47357070dea1aee9227e3a/sdk/cloudwatch/src/config.rs#L342). Does that mean that the default endpoint resolver failed to find an endpoint? I wish there were some logs if that was the case.

For the record, I'm using `eu-central-1` as Region for my tests, so it seems I should [land on monitoring.eu-central-1.amazonaws.com](https://github.com/awslabs/aws-sdk-rust/blob/31cfae2cf23be0c68a47357070dea1aee9227e3a/sdk/cloudwatch/src/aws_endpoint.rs#L6-L12)

**Reply by gagbo** [reply]
I also noticed that the headers stayed empty, even though they should get populated with the signature. I think I'll give up on using the SDK and just reimplement everything manually

**Reply by jdisanti** [reply]
Sorry this didn't work out for you. I really appreciate the time you spent digging into it though as it has revealed a lot to us. I've filed https://github.com/awslabs/smithy-rs/issues/2087 to track the issues you've run into.

I think this is a clear indication there's a bug in the default endpoint resolver when compiled for WASM. I did a quick sanity test against ARM, and saw this in the logs:
```
DEBUG aws_endpoint: resolved endpoint endpoint=Endpoint { url: "https://monitoring.us-east-1.amazonaws.com/", headers: {}, properties: ...
TRACE send_operation{operation="ListMetrics" service="cloudwatch"}: aws_smithy_http_tower::dispatch: request=Request { method: POST, uri: https://monitoring.us-east-1.amazonaws.com/, ...
```

**Reply by rcoh** [reply]
can you share the code you're using to actually make a request? I have no idea why you might be getting no endpoint from the endpoint resolver. Are you using (or replacing?) the default middleware stack? That's what will actually set the endpoint on the request.

**Reply by gagbo** [reply]
I will be able to put more info on Monday I think, but yeah, I've also been overwriting the middleware, since I don't think I could only overwrite the connector in the client builder, I think most of the relevant code goes like:

```rust
async fn main() {
    let client = new_client(Config {
        region: todo!(),
        access_key_id: todo!(),
        secret_access_key: todo!()
    }).await;
    client.list_metrics().send().await.unwrap()
}

/// Build an AWS SDK Cloudwatch client relying on runtime provided http client.
pub async fn new_client(config: Config) -> Client {
    let shared_config = config.generate_sdk_config().await;
    let config = aws_sdk_cloudwatch::config::Builder::from(&shared_config)
        .retry_config(RetryConfig::disabled())
        .build();

    let raw_client = RawBuilder::new()
        .middleware(tower::layer::util::Identity::new())
        .connector(DynConnector::new(Adapter::new()))
        .retry_config(retry::Config::default().with_max_attempts(1))
        .build();

    aws_sdk_cloudwatch::Client::with_config(
        raw_client.into_dyn_connector().into_dyn_middleware(),
        config,
    )
}

pub struct Config {
    pub region: String,
    // https://docs.rs/aws-types/0.51.0/aws_types/credentials/struct.Credentials.html#method.from_keys
    pub access_key_id: String,
    pub secret_access_key: String,
}

impl Config {
    pub async fn generate_sdk_config(&self) -> SdkConfig {
        aws_config::from_env()
            .region(Region::new(self.region.clone()))
            .credentials_provider(SharedCredentialsProvider::new(Credentials::from_keys(
                &self.access_key_id,
                &self.secret_access_key,
                None,
            )))
            .load()
            .await
    }
}

// I'm ignoring the bindings custom to my wasm situation, and our runtime isn't open-sourced yet, but
// basically having any "make_http_request" binding with the correct signature would do

// Note: downcast_error isn't added, but it's heavily inspired by the Hyper adapter in the SDK.
// Note: we are not using connector_fn because it seemed to create futures that were never woken up 
// (the provider "froze" just as it was supposed to be entering the `connector_fn`. Switching to `connector`
// solved the issue for me so I just moved on)
#[derive(Debug, Clone)]
pub struct Adapter {}

impl Service<http::Request<SdkBody>> for Adapter {
    type Response = http::Response<SdkBody>;

    type Error = ConnectorError;

    #[allow(clippy::type_complexity)]
    type Future = std::pin::Pin<
        Box<dyn std::future::Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>,
    >;

    fn poll_ready(
        &mut self,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Result<(), Self::Error>> {
        Poll::Ready(Ok(()))
    }

    fn call(&mut self, req: http::Request<SdkBody>) -> Self::Future {
        // log and make_http_request are wrappers around bindings exposed
        // by the runtime handling this wasm binary
        log(format!(
            "CloudWatch: adapter: sending request to {}",
            req.uri()
        ));
        let fut = make_http_request(self.try_into_bindings_request(req).unwrap());
        Box::pin(async move {
            Ok(http::Response::new(SdkBody::from(
                fut.await.map_err(downcast_error)?.body,
            )))
        })
    }
}
```

Unrepresented here is me using [ring's master branch](https://github.com/gagbo/aws-sdk-rust/commit/ce00a950feb5326b5450cb2358594ca535f418ed), and having a custom randomness provider that also allows to use a host-binding to act as a [custom randomness provider](https://docs.rs/getrandom/latest/getrandom/macro.register_custom_getrandom.html)

**Reply by rcoh** [reply]
ah I see—if you overwrite the middleware basically nothing will work. You want to use https://docs.rs/aws-config/0.51.0/aws_config/struct.ConfigLoader.html#method.http_connector to change the http connector—that's available in both SdkConfig and the local service config (then you don't need any other overrides).

This method was added pretty recently.

**Reply by gagbo** [reply]
I see, I got that part working I think, I'll still be using a custom-made API wrapper: it seems that the default middleware uses standard library time to get the date (which is expected), but that also means that on wasm targets (probably without bindgen), that leads to a panic `CloudWatch panic: panicked at 'time not implemented on this platform', library/std/src/sys/wasm/../unsupported/time.rs:31:9`.

At least, for the record, here's what I ended up with:
```rust
use aws_sdk_cloudwatch::Client;
/// Build an AWS SDK Cloudwatch client relying on runtime provided http client.
pub async fn new_client(config: Config) -> Client {
    let shared_config = config.generate_sdk_config().await;
    let config = aws_sdk_cloudwatch::config::Builder::from(&shared_config)
        .retry_config(RetryConfig::disabled())
        .build();

    Client::from_conf_conn(config, Adapter::new())
}
```

**Reply by jdisanti** [reply]
I suspect it's grabbing the time to create the SigV4 signature. Currently, we have a `AsyncSleep` trait for making the async sleep implementation configurable. I'm now thinking maybe we should rename that to `AsyncTimer` or something, and give it the ability to both sleep and tell time so that this use-case is supportable.

**Reply by amitu** [reply]
@gagbo I am also going to try to use aws sdk in wasm environment (`wasm32-unknown-unknown`), how did it go for you eventually?

**Reply by gagbo** [reply, marked as answer]
I rewrote the deserialization/serialization/signature from scratch on every endpoint I wanted to use, see for example https://github.com/fiberplane/providers/blob/main/providers/cloudwatch/src/client/cloudwatch.rs and https://github.com/fiberplane/providers/blob/main/providers/cloudwatch/src/client/canonical_request.rs

Note that since then, I thought the AWS SDK got rid of the version of `ring` that was triggering the compilation issues for the `wasm32-unknown-unknown` targets. Maybe just compiling will work

EDIT: Or rather. as mentionned above, that a trait was created (`AsyncTimer` but it might have changed name) that allows to inject your time function into the SDK
